File: src/com/strobecorp/kirk/Deploy.java
package com.strobecorp.kirk;

import java.io.IOException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.ServletException;
import org.eclipse.jetty.server.Request;
import org.jruby.embed.PathType;
import org.jruby.embed.LocalContextScope;
import org.jruby.embed.ScriptingContainer;
import org.jruby.runtime.builtin.IRubyObject;

public class Deploy {

  private ApplicationConfig  config;
  private ScriptingContainer context;
  private Object             bootstrapper;
  private Object             handler;
  private String             key;

  public Deploy(ApplicationConfig config) {
    this.config = config;
    initializeScriptingContext();
  }

  public void prepare() {
    this.handler = context.callMethod(bootstrapper, "run",
      config.getRackupPath());
  }

  public void handle(String target, Request baseRequest,
      HttpServletRequest request, HttpServletResponse response)
      throws IOException, ServletException {
    context.callMethod(handler, "handle", target, baseRequest, request, response);
  }

  public String getKey() {
    return key;
  }

  public void terminate() {
    this.context.terminate();
    this.context = null;
    this.handler = null;
  }

  private void initializeScriptingContext() {
    this.context = new ScriptingContainer(LocalContextScope.THREADSAFE);
    this.bootstrapper = context.runScriptlet(
      PathType.ABSOLUTE, config.getBootstrapPath());

    key = (String) context.callMethod(bootstrapper, "warmup",
      config.getApplicationPath());
  }

}

File: lib/kirk/deploy.rb
require 'rack'
module Kirk
  class Deploy
    def initialize(app)
      @app = app
      @context, @handler = spawn_context
    end

    def handle(target, base_request, request, response)
      @handler.handle(target, base_request, request, response)
    end

    def terminate
      @context.terminate
    end

  private

    def spawn_context
      context = ScriptingContainer.new(LocalContextScope::THREADSAFE)
      handler = context.run_scriptlet(bootstrap_script)

      [ context, handler ]
    end

    def bootstrap_script
      <<-RUBY
        # Change into the directory
        Dir.chdir "#{application_path}"

        # Load the Gemfile
        require "rubygems"
        require "bundler/setup"

        # Load the Kirk handler
        $:.unshift "#{runtime_path}"
        require "kirk/handler"

        Kirk::Handler.new
      RUBY
    end

    def runtime_path
      File.dirname(File.dirname(__FILE__))
    end

    def application_path
      @app.path
    end
  end
end
